_posts/2020-06-23-user-feedback-websockets.md

```diff
@@ -0,0 +1,289 @@
+mqtt.default_pass = $RABBITMQ_DEFAULT_PASS  
+mqtt.allow_anonymous  = true  
+
+## Subscribing with a JS client
+
+[Eclipse](https://www.eclipse.org/paho/clients/js/){:target="_blank" rel="noopener noreferrer"} offers us a JS client library to use for opening a WebSocket over MQTT  
+
+With some basic setup, we can fix ourselves a quick WebSocket to the Rabbit to test the handshake  
+As the JS client, we will be subscribing to a queue and listen for any notifications from the backend.   
+To configure our client, we need to know what properties we need.  
+A list of properties can be found in the [documentation](https://www.eclipse.org/paho/files/jsdoc/Paho.MQTT.Client.html){:target="_blank" rel="noopener noreferrer"}.  
+The most important ones are enabling SSL and using the keep-alive period as described above.  
+
+````javascript
+var wsbroker = '{rabbitmq_hostname/ws}'
+var wsport = 443; // port for above
+var client = new Paho.MQTT.Client(wsbroker,wsport, "?access_token=eyJ0eX","{client}"); // you can use randomizer to be unique "myclientid_" + parseInt(Math.random() * 100, 10));
+    
+client.onConnectionLost = function (responseObject) {
+    console.log("CONNECTION LOST - " + responseObject.errorMessage);
+};
+client.onMessageArrived = function (message) {
+    console.log("RECEIVE ON " + message.destinationName + " PAYLOAD " + message.payloadString);
+};
+  
+  
+client.connect({
+    useSSL: true,
+    onSuccess: function () {
+        console.log("CONNECTION SUCCESS");
+        client.subscribe('events', {qos: 0});
+    },
+    onFailure: function (message) {
+        debug("CONNECTION FAILURE - " + message.errorMessage);
+    }
+});
+````
+### Keeping the heartbeat alive
+At any point after the handshake, either the client or the server can choose to send a ping to the other party.  
+When the ping is received, the recipient must send back a pong as soon as possible.  
+You can use this to make sure the client is still connected.  
+A best practice is to set the heartbeat between 20-30 seconds [https://tools.ietf.org/html/rfc6202#page-13](https://tools.ietf.org/html/rfc6202#page-13){:target="_blank" rel="noopener noreferrer"}  
+The client sends a ping every 10 seconds, and the server waits 10 seconds to send back a pong.  
+
+
+#### MQTT keep-alive period
+The keep-alive period is the answer from the MQTT protocol to the WebSocket heartbeat.  
+The keep-alive is a time interval measured in seconds, expressed as a 16-bit word, it is the maximum time interval that is permitted to elapse between the point at which the client finishes transmitting one control package and the point it starts sending the next.  
+It is the responsibility of the client to ensure the interval between the control packets being sent does not exceed the keep-alive value.  
+The client can send a PING at any time, irrespective of the keep-alive value, and use the PONG to determine that the network and the server are working.  
```

Comment:
be consistent, you used ping and pong before, yet here you use PING and PONG. Pick one ;)